native/ios/WatermelonDB/Database.cpp

```diff
@@ -1 +1,533 @@
+    int count = sqlite3_column_int(statement.stmt, 0);
+
+    return jsi::Value(count);
+}
+
+void Database::batch(jsi::Array &operations) {
+    auto &rt = getRt();
+    beginTransaction();
+
+    std::vector<std::string> addedIds = {};
+    std::vector<std::string> removedIds = {};
+
+    try {
+        size_t operationsCount = operations.length(rt);
+        for (size_t i = 0; i < operationsCount; i++) {
+            jsi::Array operation = operations.getValueAtIndex(rt, i).getObject(rt).getArray(rt);
+            std::string type = operation.getValueAtIndex(rt, 0).getString(rt).utf8(rt);
+            const jsi::String table = operation.getValueAtIndex(rt, 1).getString(rt);
+
+            if (type == "create") {
+                std::string id = operation.getValueAtIndex(rt, 2).getString(rt).utf8(rt);
+                std::string sql = operation.getValueAtIndex(rt, 3).getString(rt).utf8(rt);
+                jsi::Array arguments = operation.getValueAtIndex(rt, 4).getObject(rt).getArray(rt);
+
+                executeUpdate(sql, arguments);
+                addedIds.push_back(cacheKey(table.utf8(rt), id));
+            } else if (type == "execute") {
+                jsi::String sql = operation.getValueAtIndex(rt, 2).getString(rt);
+                jsi::Array arguments = operation.getValueAtIndex(rt, 3).getObject(rt).getArray(rt);
+
+                executeUpdate(sql.utf8(rt), arguments);
+            } else if (type == "markAsDeleted") {
+                const jsi::String id = operation.getValueAtIndex(rt, 2).getString(rt);
+                auto args = jsi::Array::createWithElements(rt, id);
+                executeUpdate("update " + table.utf8(rt) + " set _status='deleted' where id == ?", args);
+
+                removedIds.push_back(cacheKey(table.utf8(rt), id.utf8(rt)));
+            } else if (type == "destroyPermanently") {
+                const jsi::String id = operation.getValueAtIndex(rt, 2).getString(rt);
+                auto args = jsi::Array::createWithElements(rt, id);
+
+                // TODO: What's the behavior if nothing got deleted?
+                executeUpdate("delete from " + table.utf8(rt) + " where id == ?", args);
+                removedIds.push_back(cacheKey(table.utf8(rt), id.utf8(rt)));
+            } else {
+                throw jsi::JSError(rt, "Invalid operation type");
+            }
+        }
+        commit();
+    } catch (const std::exception &ex) {
+        rollback();
+        throw;
+    }
+
+    for (auto const &key : addedIds) {
+        markAsCached(key);
+    }
+
+    for (auto const &key : removedIds) {
+        removeFromCache(key);
+    }
+}
+
+jsi::Array Database::getDeletedRecords(jsi::String &tableName) {
+    auto &rt = getRt();
+    auto args = jsi::Array::createWithElements(rt);
+    auto statement = executeQuery("select id from " + tableName.utf8(rt) + " where _status='deleted'", args);
+
+    jsi::Array records(rt, 0);
+
+    for (size_t i = 0; true; i++) {
```

Comment:
is it guaranteed that `stepResult` eventually will be equal to `SQLITE_DONE`? Otherwise you could end with infinite loop

Reply:
yep - we get SQLITE_ROW zero or more (but not infinite) number of times, followed by SQLITE_DONE — or another status for error (which will break out of the loop by throwing).